digdag-core/src/main/java/io/digdag/core/agent/ExtractArchiveWorkspaceManager.java

```diff
@@ -22,19 +31,53 @@
     public ExtractArchiveWorkspaceManager(TempFileManager tempFiles)
     {
         this.tempFiles = tempFiles;
+
     }
 
     @Override
     public <T> T withExtractedArchive(TaskRequest request, ArchiveProvider archiveProvider, WithWorkspaceAction<T> func)
             throws IOException
     {
-        try (TempDir workspacePath = createNewWorkspace(request)) {
-            Optional<StorageObject> in = archiveProvider.open();
-            if (in.isPresent()) {
-                ProjectArchives.extractTarArchive(workspacePath.get(), in.get().getContentInputStream());
+        TempDir workspacePath = null;
+        try {
+            try {
+                // A temporary workspace should be created for each try
+                // And always close the temporary workspace
+                workspacePath = RetryExecutor.retryExecutor()
+                        .retryIf(exception -> true)
+                        .withInitialRetryWait(EXTRACT_MIN_RETRY_WAIT_MS)
+                        .withMaxRetryWait(EXTRACT_MAX_RETRY_WAIT_MS)
+                        .onRetry((exception, retryCount, retryLimit, retryWait) ->
+                                logger.warn("Failed to extract archive: retry {} of {}", retryCount, retryLimit, exception))
+                        .withRetryLimit(EXTRACT_RETRIES)
+                        .run(() -> {
+                            TempDir newWorkSpacePath = null;
+                            try {
+                                newWorkSpacePath = createNewWorkspace(request);
+                                Optional<StorageObject> in = archiveProvider.open();
+                                if (in.isPresent()) {
+                                    ProjectArchives.extractTarArchive(newWorkSpacePath.get(), in.get().getContentInputStream());
+                                }
+                                return newWorkSpacePath;
+                            }
+                            catch (Throwable e) {
+                                if (newWorkSpacePath != null) {
+                                    newWorkSpacePath.close();
+                                }
+                                throw e;
+                            }
+                        });
+            }
+            catch (RetryExecutor.RetryGiveupException e) {
+                Throwables.propagate(e);
```

Comment:
I think `Throwables.propagate(e.getCause())` is better just to cleanup stacktrace in log messages.

And `Throwables.propagate` is marked as deprecated. I guess that it's because we often miss `throw` as you also did...Throwables.propagate returns a Throwable and doesn't throw.

So we should use `Throwables.throwIfUnchecked` like this instead:

```java
Throwables.throwIfUnchecked(e.getCause());
throw new RuntimeException(e.getCause());
```

Reply:
`Throwables` has these implementations as of that throw an exception not just returning an exception object and I can't find `@Deprecated` annotation on `propagate`. Also, I can't find `Throwables.throwIfUnchecked`

```
  public static RuntimeException propagate(Throwable throwable) {
    propagateIfPossible(checkNotNull(throwable));
    throw new RuntimeException(throwable);
  }

  public static void propagateIfPossible(@Nullable Throwable throwable) {
    propagateIfInstanceOf(throwable, Error.class);
    propagateIfInstanceOf(throwable, RuntimeException.class);
  }

  public static <X extends Throwable> void propagateIfInstanceOf(
      @Nullable Throwable throwable, Class<X> declaredType) throws X {
    // Check for null is needed to avoid frequent JNI calls to isInstance().
    if (throwable != null && declaredType.isInstance(throwable)) {
      throw declaredType.cast(throwable);
    }
  }
```

So, I think just `Throwables.propagate(e.getCause())` is enough.